lib/value/src/kind/insert.rs

```diff
@@ -0,0 +1,661 @@
+//! All types related to inserting one [`Kind`] into another.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use crate::kind::merge;
+
+use super::Kind;
+use lookup::{Field, Lookup, Segment};
+
+/// The strategy to use when an inner segment in a path does not match the actual `Kind`
+/// present.
+///
+/// For example, if a path expects an object with a given field at a certain path segment,
+/// but the `Kind` defines there to be a non-object type, then one of the below actions
+/// must be taken to succeed (or fail) at inserting the provided `Kind` at its final
+/// destination.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum InnerConflict {
+    /// Keep the existing `Kind` states, but add the required state (object or array) to
+    /// accomodate the final `path` structure.
+    Merge(merge::Strategy),
+
+    /// Remove the existing `Kind` states, replacing it for a singular object or array
+    /// `Kind` state.
+    Replace,
+
+    /// Reject the insertion, returning an error.
+    Reject,
+}
+
+/// The strategy to use when the leaf segment already has a `Kind` present.
```

Comment:
Leaf segment == final path segment?

Reply:
Correct. It's common terminology for tree-like structures, although paths aren't necessarily a tree structure, so if you think it's confusing, let me know and we can change it.

Reply:
Nope - leaf is fine, just wanted to double check as I mid brain melt at this point :)